### PR TITLE
Add c unity package

### PR DESCRIPTION
**Pull request by fwqaaq**
* Simple Unit Testing for C: <https://github.com/ThrowTheSwitch/Unity>

### CODE REVIEW

**Comment by waruqi** [comment]
please push to dev branch next

File: packages/u/unity/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+    add_urls("https://github.com/ThrowTheSwitch/Unity/archive/refs/tags/$(version).tar.gz", "https://github.com/ThrowTheSwitch/Unity")
+
+    add_versions("v2.5.2", "3786de6c8f389be3894feae4f7d8680a02e70ed4dbcce36109c8f8646da2671a")
+    add_versions("v2.5.1", "5ce08ef62f5f64d18f8137b3eaa6d29199ee81d1fc952cef0eea96660a2caf47")
+    add_versions("v2.5.0", "d470165dc46652cf73fda54e2650b483b94a7e24ccf793ded28615729b2f41ed")
+    add_versions("v2.4.3", "a8c5e384f511a03c603bbecc9edc24d2cb4a916998d51a29cf2e3a2896920d03")
+
+    on_install(function (package)
+        os.cp("src/*.c", package:installdir("lib"))
```
Comment (waruqi) [inline packages/u/unity/xmake.lua:14]: why?

File: packages/u/unity/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package("unity")
+    set_homepage("http://www.throwtheswitch.org/unity")
+    set_description("Simple Unit Testing for C")
+    set_license("MIT")
+
+    add_urls("https://github.com/ThrowTheSwitch/Unity/archive/refs/tags/$(version).tar.gz", "https://github.com/ThrowTheSwitch/Unity")
```
Comment (waruqi) [inline packages/u/unity/xmake.lua:6]: add a newline and use git url `https://github.com/ThrowTheSwitch/Unity.git`

```lua
add_urls("https://github.com/ThrowTheSwitch/Unity/archive/refs/tags/$(version).tar.gz", 
         "https://github.com/ThrowTheSwitch/Unity.git")
```

File: packages/u/unity/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package("unity")
```
Comment (waruqi) [inline packages/u/unity/xmake.lua:1]: please rename to unity_test

File: packages/u/unity/xmake.lua (hunk 1 above)
Comment (fwqaaq) [reply packages/u/unity/xmake.lua:14]: It's only a [`unity.c`](https://github.com/ThrowTheSwitch/Unity/tree/master/src) lib, maybe I should change to `unity.c`? Or Don't we need lib files?

File: packages/u/unity/xmake.lua (hunk 2 above)
Comment (fwqaaq) [reply packages/u/unity/xmake.lua:6]: OK

File: packages/u/unity/xmake.lua (hunk 3 above)
Comment (fwqaaq) [reply packages/u/unity/xmake.lua:1]: Thanks

File: packages/u/unity/xmake.lua (hunk 1 above)
Comment (waruqi) [reply packages/u/unity/xmake.lua:14]: meson and cmake will compile it as lib. https://github.com/ThrowTheSwitch/Unity/blob/1b9199ee380e203603b6649df9510db9cab147d9/CMakeLists.txt#L69

Comment (fwqaaq) [reply packages/u/unity/xmake.lua:14]: So, xmake will help me do anything, I onlye need to add a `.h` file?

Comment (Qudix) [reply packages/u/unity/xmake.lua:14]: This isn't a header-only library, it needs to be compiled and linked to.

Comment (fwqaaq) [reply packages/u/unity/xmake.lua:14]: oh, I'm newbie, I have to look at the document again

**Comment by fwqaaq** [comment]
@waruqi Ohhh, it always failed when I run test specifying the windows platform. (:

```bash
xmake l scripts/test.lua --shallow -vD -k shared -p windows unity_test 
```

I just finished building with the ninja, so haven't commited. And these `action tests` also haven't windows platform?

**Comment by waruqi** [comment]
> @waruqi Ohhh, it always failed when I run test specifying the windows platform. (:
> 
> ```shell
> xmake l scripts/test.lua --shallow -vD -k shared -p windows unity_test 
> ```
> 
> I just finished building with the ninja, so haven't commited.

it works now on windows ci

**Comment by fwqaaq** [comment]
> it works now on windows ci

Yeah, yeah, but no successed on my linux

**Comment by fwqaaq** [comment]
I have to try your code, I don't know why, 😢

**Comment by waruqi** [comment]
> > it works now on windows ci
> 
> Yeah, yeah, but no successed on my linux

but it works on linux ci

**Comment by fwqaaq** [comment]
```
error: @programdir/modules/package/tools/cmake.lua:89: vs not found!
stack traceback:
    [C]: in function 'error'
    [@programdir/core/base/os.lua:913]: in function 'raiselevel'
    [@programdir/core/sandbox/modules/utils.lua:149]: in function 'assert'
    [@programdir/modules/package/tools/cmake.lua:89]: in function '_get_msvc'
    [@programdir/modules/package/tools/cmake.lua:621]: in function '_get_cmake_generator_for_msvc'
    [@programdir/modules/package/tools/cmake.lua:654]: in function '_get_configs_for_generator'
    [@programdir/modules/package/tools/cmake.lua:687]: in function '_get_configs'
    [@programdir/modules/package/tools/cmake.lua:980]: in function 'install'
    [...feiwu/Project/xmake-repo/packages/u/unity_test/xmake.lua:20]: in function 'script'
    [...dir/modules/private/action/require/impl/utils/filter.lua:125]: in function 'call'
    [.../modules/private/action/require/impl/actions/install.lua:364]:

  => install unity_test v2.5.2 .. failed
error: @programdir/modules/private/async/runjobs.lua:256: .../modules/private/action/require/impl/actions/install.lua:469: install failed!
stack traceback:
    [C]: in function 'error'
    [@programdir/core/base/os.lua:913]:
    [.../modules/private/action/require/impl/actions/install.lua:469]: in function 'catch'
    [@programdir/core/sandbox/modules/try.lua:123]: in function 'try'
    [.../modules/private/action/require/impl/actions/install.lua:328]:
    [...modules/private/action/require/impl/install_packages.lua:479]: in function 'jobfunc'
    [@programdir/modules/private/async/runjobs.lua:232]:

stack traceback:
        [C]: in function 'error'
        @programdir/core/base/os.lua:913: in function 'base/os.raiselevel'
        (...tail calls...)
        @programdir/modules/private/async/runjobs.lua:256: in field 'catch'
        @programdir/core/sandbox/modules/try.lua:123: in global 'try'
        @programdir/modules/private/async/runjobs.lua:218: in upvalue 'cotask'
        @programdir/core/base/scheduler.lua:404: in function <@programdir/core/base/scheduler.lua:397>
error: execv(xmake require -f -y --build -v -D --shallow --extra={configs={shared=true}} unity_test) failed(255)
```

But I ran into this issue that `vs not found`, I used your code as well.

**Comment by waruqi** [comment]
> ```
> error: @programdir/modules/package/tools/cmake.lua:89: vs not found!
> stack traceback:
>     [C]: in function 'error'
>     [@programdir/core/base/os.lua:913]: in function 'raiselevel'
>     [@programdir/core/sandbox/modules/utils.lua:149]: in function 'assert'
>     [@programdir/modules/package/tools/cmake.lua:89]: in function '_get_msvc'
>     [@programdir/modules/package/tools/cmake.lua:621]: in function '_get_cmake_generator_for_msvc'
>     [@programdir/modules/package/tools/cmake.lua:654]: in function '_get_configs_for_generator'
>     [@programdir/modules/package/tools/cmake.lua:687]: in function '_get_configs'
>     [@programdir/modules/package/tools/cmake.lua:980]: in function 'install'
>     [...feiwu/Project/xmake-repo/packages/u/unity_test/xmake.lua:20]: in function 'script'
>     [...dir/modules/private/action/require/impl/utils/filter.lua:125]: in function 'call'
>     [.../modules/private/action/require/impl/actions/install.lua:364]:
> 
>   => install unity_test v2.5.2 .. failed
> error: @programdir/modules/private/async/runjobs.lua:256: .../modules/private/action/require/impl/actions/install.lua:469: install failed!
> stack traceback:
>     [C]: in function 'error'
>     [@programdir/core/base/os.lua:913]:
>     [.../modules/private/action/require/impl/actions/install.lua:469]: in function 'catch'
>     [@programdir/core/sandbox/modules/try.lua:123]: in function 'try'
>     [.../modules/private/action/require/impl/actions/install.lua:328]:
>     [...modules/private/action/require/impl/install_packages.lua:479]: in function 'jobfunc'
>     [@programdir/modules/private/async/runjobs.lua:232]:
> 
> stack traceback:
>         [C]: in function 'error'
>         @programdir/core/base/os.lua:913: in function 'base/os.raiselevel'
>         (...tail calls...)
>         @programdir/modules/private/async/runjobs.lua:256: in field 'catch'
>         @programdir/core/sandbox/modules/try.lua:123: in global 'try'
>         @programdir/modules/private/async/runjobs.lua:218: in upvalue 'cotask'
>         @programdir/core/base/scheduler.lua:404: in function <@programdir/core/base/scheduler.lua:397>
> error: execv(xmake require -f -y --build -v -D --shallow --extra={configs={shared=true}} unity_test) failed(255)
> ```
> 
> But I ran into this issue that `vs not found`, I used your code as well.

vs? windows? did you install vs?

**Comment by fwqaaq** [comment]
> vs? windows? did you install vs?

Debian, bro!

```
 uname -a                                                                                                                                                          
Linux feiwu 6.1.0-10-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.38-1 (2023-07-14) x86_64 GNU/Linux
```

**Comment by waruqi** [comment]
> > ```
> > error: @programdir/modules/package/tools/cmake.lua:89: vs not found!
> > stack traceback:
> >     [C]: in function 'error'
> >     [@programdir/core/base/os.lua:913]: in function 'raiselevel'
> >     [@programdir/core/sandbox/modules/utils.lua:149]: in function 'assert'
> >     [@programdir/modules/package/tools/cmake.lua:89]: in function '_get_msvc'
> >     [@programdir/modules/package/tools/cmake.lua:621]: in function '_get_cmake_generator_for_msvc'
> >     [@programdir/modules/package/tools/cmake.lua:654]: in function '_get_configs_for_generator'
> >     [@programdir/modules/package/tools/cmake.lua:687]: in function '_get_configs'
> >     [@programdir/modules/package/tools/cmake.lua:980]: in function 'install'
> >     [...feiwu/Project/xmake-repo/packages/u/unity_test/xmake.lua:20]: in function 'script'
> >     [...dir/modules/private/action/require/impl/utils/filter.lua:125]: in function 'call'
> >     [.../modules/private/action/require/impl/actions/install.lua:364]:
> > 
> >   => install unity_test v2.5.2 .. failed
> > error: @programdir/modules/private/async/runjobs.lua:256: .../modules/private/action/require/impl/actions/install.lua:469: install failed!
> > stack traceback:
> >     [C]: in function 'error'
> >     [@programdir/core/base/os.lua:913]:
> >     [.../modules/private/action/require/impl/actions/install.lua:469]: in function 'catch'
> >     [@programdir/core/sandbox/modules/try.lua:123]: in function 'try'
> >     [.../modules/private/action/require/impl/actions/install.lua:328]:
> >     [...modules/private/action/require/impl/install_packages.lua:479]: in function 'jobfunc'
> >     [@programdir/modules/private/async/runjobs.lua:232]:
> > 
> > stack traceback:
> >         [C]: in function 'error'
> >         @programdir/core/base/os.lua:913: in function 'base/os.raiselevel'
> >         (...tail calls...)
> >         @programdir/modules/private/async/runjobs.lua:256: in field 'catch'
> >         @programdir/core/sandbox/modules/try.lua:123: in global 'try'
> >         @programdir/modules/private/async/runjobs.lua:218: in upvalue 'cotask'
> >         @programdir/core/base/scheduler.lua:404: in function <@programdir/core/base/scheduler.lua:397>
> > error: execv(xmake require -f -y --build -v -D --shallow --extra={configs={shared=true}} unity_test) failed(255)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But I ran into this issue that `vs not found`, I used your code as well.
> 
> vs? windows? did you install vs?

what's you xmake version? did you modify xmake source code?

**Comment by fwqaaq** [comment]
> what's your xmake version? did you modify xmake source code?

```
xmake --version                                                                                                                                               
xmake v2.8.1+20230723, A cross-platform build utility based on Lua
```

no modified

**Comment by waruqi** [comment]
> ```shell
>  -p windows 
> ```

did you pass ` -p windows ` to test.lua? it's only for windows platform.

**Comment by fwqaaq** [comment]
> did you pass -p windows to test.lua? it's only for windows platform.

Is it like that? But successed when I used `ninja` to build `-p macosx`. Or only `-p windows` is like that?

**Comment by waruqi** [comment]
> > did you pass -p windows to test.lua? it's only for windows platform.
> 
> Is it like that? But successed when I used `ninja` to build `-p macosx`. Or only `-p windows` is like that?

? if you run test.lua on linux, you cannot pass `-p windows`.

```
xmake l scripts/test.lua --shallow unity_test
```

**Comment by fwqaaq** [comment]
> ? if you run test.lua on linux, you cannot pass -p windows.

Oh I see, I thought that this is a cross platroms test before. So Only using `xmake l scripts/test.lua --shallow unity_test` is Ok?

**Comment by waruqi** [comment]
> > ? if you run test.lua on linux, you cannot pass -p windows.
> 
> Oh I see, I thought that this is a cross platroms test before. So Only using `xmake l scripts/test.lua --shallow unity_test` is Ok?

right, msvc has no cross-platform toolchain. you should use mingw.

**Pull request merged**